File: pythonOOP/solutions/solution_02_class_methods_static.py
"""
SOLUTION 2: Class Methods and Static Methods
=============================================

CONCEPTS EXPLAINED:
-------------------
1. Class variables: Shared by all instances of the class
2. Instance variables: Unique to each instance
3. @classmethod: Works with the class itself, not instances
4. @staticmethod: Doesn't access class or instance data
5. cls: Refers to the class (like self refers to instance)

"""

class Employee:
    """Manages employee records for a company"""
    
    # Class variables - shared by ALL employees
    company_name = "TechCorp"
    employee_count = 0
    
    def __init__(self, name, position, salary):
        """
        Initialize an Employee object
        
        Parameters:
        -----------
        name : str
            Employee's name
        position : str
            Job position
        salary : float
            Annual salary
        """
        # Instance variables - unique to each employee
        self.name = name
        self.position = position
        self.salary = salary
        
        # Increment class variable and assign employee ID
        Employee.employee_count += 1
        self.employee_id = Employee.employee_count
    
    @classmethod
    def get_employee_count(cls):
        """
        Class method to get total number of employees
        
        Returns:
        --------
        int : Total number of employees
        
        NOTE: Use @classmethod when you need to access or modify class variables
        'cls' refers to the class itself (Employee)
        """
        return cls.employee_count
    
    @classmethod
    def set_company_name(cls, new_name):
        """
        Class method to update company name
        
        Parameters:
        -----------
        new_name : str
            New company name
        
        NOTE: This changes the company name for ALL employees
        """
        cls.company_name = new_name
        print(f"Company name updated to: {cls.company_name}")
    
    @staticmethod
    def is_valid_salary(salary):
        """
        Static method to validate salary range
        
        Parameters:
        -----------
        salary : float
            Salary amount to validate
        
        Returns:
        --------
        bool : True if salary is valid, False otherwise
        
        NOTE: Use @staticmethod when the method doesn't need access to
        class or instance data. It's like a regular function but
        belongs logically to the class.
        """
        return 20000 <= salary <= 500000
    
    def give_raise(self, percentage):
        """
        Increase employee's salary by given percentage
        
        Parameters:
        -----------
        percentage : float
            Percentage increase (e.g., 10 for 10%)
        """
        if percentage < 0:
            print("Error: Percentage cannot be negative")
            return
        
        old_salary = self.salary
        self.salary = self.salary * (1 + percentage / 100)
        print(f"Salary increased from ${old_salary:,.2f} to ${self.salary:,.2f}")
    
    def display_info(self):
        """Display all employee information"""
        print(f"\n{'=' * 50}")
        print(f"Company: {Employee.company_name}")
        print(f"Employee ID: {self.employee_id}")
        print(f"Name: {self.name}")
        print(f"Position: {self.position}")
        print(f"Salary: ${self.salary:,.2f}")
        print(f"{'=' * 50}")


# ============================================
# TESTING THE CODE
# ============================================

if __name__ == "__main__":
    print("=" * 60)
    print("TESTING EMPLOYEE CLASS")
    print("=" * 60)
    
    # Test static method before creating any employees
    print("\n1. Testing static method (salary validation):")
    print(f"Is $50,000 valid? {Employee.is_valid_salary(50000)}")  # True
    print(f"Is $15,000 valid? {Employee.is_valid_salary(15000)}")  # False
    print(f"Is $600,000 valid? {Employee.is_valid_salary(600000)}")  # False
    
    # Creating employee objects
    print("\n2. Creating employees:")
    emp1 = Employee("Alice Johnson", "Software Engineer", 85000)
    emp2 = Employee("Bob Smith", "Data Scientist", 95000)
    emp3 = Employee("Carol White", "Product Manager", 105000)
    
    # Display employee information
    print("\n3. Employee Details:")
    emp1.display_info()
    emp2.display_info()
    emp3.display_info()
    
    # Test class method - get employee count
    print("\n4. Testing class method (get employee count):")
    print(f"Total employees: {Employee.get_employee_count()}")
    
    # Test class method - set company name
    print("\n5. Testing class method (set company name):")
    Employee.set_company_name("InnovateTech Solutions")
    
    # Notice company name changed for all employees
    print("\n6. After company name change:")
    emp1.display_info()
    
    # Test give_raise method
    print("\n7. Testing give_raise():")
    print(f"\nBefore raise - {emp1.name}'s salary: ${emp1.salary:,.2f}")
    emp1.give_raise(10)  # 10% raise
    print(f"After raise - {emp1.name}'s salary: ${emp1.salary:,.2f}")
    
    # Create more employees to test employee_count
    print("\n8. Creating more employees:")
    emp4 = Employee("David Brown", "DevOps Engineer", 90000)
    print(f"Total employees now: {Employee.get_employee_count()}")
    
    # Access class variables
    print("\n9. Accessing class variables:")
    print(f"Via class: {Employee.company_name}")
    print(f"Via instance: {emp1.company_name}")
    print(f"Employee count: {Employee.employee_count}")
    
    print("\n" + "=" * 60)
    print("KEY TAKEAWAYS:")
    print("=" * 60)
    print("1. Class variables are shared among all instances")
    print("2. @classmethod receives the class (cls) as first parameter")
    print("3. @staticmethod doesn't receive class or instance")
    print("4. Class methods can modify class state")
    print("5. Static methods are utility functions within the class")
    print("6. Access class variables via ClassName.variable")

File: javaOOP/solutions/Solution02ClassMethodsStatic.java
/**
 * SOLUTION 2: Static Methods and Variables
 * =========================================
 * 
 * CONCEPTS EXPLAINED:
 * -------------------
 * 1. Static variables: Shared among all instances of a class
 * 2. Static methods: Can be called without creating an object
 * 3. Instance variables: Unique to each object
 * 4. Instance methods: Called on specific objects
 * 
 * KEY TAKEAWAYS:
 * --------------
 * - Static members belong to the class, not to any specific instance
 * - Static methods can only access static members directly
 * - Instance methods can access both static and instance members
 * - Use static for shared data (like counters, constants)
 * - Use static methods for utility functions that don't need object state
 */

package solutions;

class Employee {
    // STATIC VARIABLES - shared by all Employee instances
    private static String companyName = "TechCorp";
    private static int employeeCount = 0;
    private static int nextId = 1;
    
    // INSTANCE VARIABLES - unique to each Employee object
    private String name;
    private int id;
    private double salary;
    
    /**
     * Constructor to create an Employee
     * Automatically assigns ID and increments counters
     * 
     * @param name Employee name
     * @param salary Employee salary
     */
    public Employee(String name, double salary) {
        this.name = name;
        this.salary = salary;
        this.id = nextId++;        // Assign current ID and increment for next
        employeeCount++;            // Increment total employee count
    }
    
    /**
     * Display information about this specific employee (instance method)
     */
    public void displayInfo() {
        System.out.println("\nEmployee #" + id);
        System.out.println("Name: " + name);
        System.out.println("Salary: $" + String.format("%.2f", salary));
        System.out.println("Company: " + companyName);
    }
    
    /**
     * Give a raise to this employee (instance method)
     * 
     * @param percentage Percentage increase (e.g., 10 for 10%)
     */
    public void giveRaise(double percentage) {
        if (percentage > 0) {
            double oldSalary = salary;
            salary = salary * (1 + percentage / 100);
            System.out.println("✓ " + name + " received a " + percentage + "% raise!");
            System.out.println("  Salary: $" + String.format("%.2f", oldSalary) + 
                             " → $" + String.format("%.2f", salary));
        } else {
            System.out.println("✗ Invalid raise percentage");
        }
    }
    
    // STATIC METHODS - can be called without creating an Employee object
    
    /**
     * Set the company name for all employees
     * 
     * @param name New company name
     */
    public static void setCompanyName(String name) {
        companyName = name;
        System.out.println("✓ Company name set to: " + companyName);
    }
    
    /**
     * Get the total number of employees created
     * 
     * @return Total employee count
     */
    public static int getEmployeeCount() {
        return employeeCount;
    }
    
    /**
     * Display information about the company
     */
    public static void displayCompanyInfo() {
        System.out.println("\n" + "=".repeat(40));
        System.out.println("COMPANY INFORMATION");
        System.out.println("=".repeat(40));
        System.out.println("Company Name: " + companyName);
        System.out.println("Total Employees: " + employeeCount);
        System.out.println("Next Employee ID: " + nextId);
        System.out.println("=".repeat(40));
    }
    
    // Getters
    public String getName() { return name; }
    public int getId() { return id; }
    public double getSalary() { return salary; }
    public static String getCompanyName() { return companyName; }
}

/**
 * Main class to test the Employee class
 */
public class Solution02ClassMethodsStatic {
    public static void main(String[] args) {
        System.out.println("=".repeat(50));
        System.out.println("TESTING EMPLOYEE CLASS - STATIC vs INSTANCE");
        System.out.println("=".repeat(50));
        
        // STATIC METHOD - called without creating an object
        System.out.println("\n1. Setting company name (static method):");
        Employee.setCompanyName("InnovateTech Solutions");
        
        // STATIC METHOD - get employee count before creating any employees
        System.out.println("\n2. Initial employee count: " + Employee.getEmployeeCount());
        
        // Creating employee objects
        System.out.println("\n3. Creating employees:");
        Employee emp1 = new Employee("Alice Johnson", 75000);
        System.out.println("✓ Created: " + emp1.getName());
        
        Employee emp2 = new Employee("Bob Smith", 80000);
        System.out.println("✓ Created: " + emp2.getName());
        
        Employee emp3 = new Employee("Carol Williams", 85000);
        System.out.println("✓ Created: " + emp3.getName());
        
        // STATIC METHOD - display company info
        System.out.println("\n4. Company information:");
        Employee.displayCompanyInfo();
        
        // INSTANCE METHODS - called on specific objects
        System.out.println("\n5. Employee details:");
        emp1.displayInfo();
        emp2.displayInfo();
        emp3.displayInfo();
        
        // Give raises
        System.out.println("\n6. Giving raises:");
        emp1.giveRaise(10);    // 10% raise for Alice
        emp2.giveRaise(15);    // 15% raise for Bob
        
        // Display updated info
        System.out.println("\n7. Updated employee info:");
        emp1.displayInfo();
        emp2.displayInfo();
        
        // Demonstrate static variable sharing
        System.out.println("\n8. Demonstrating static variable sharing:");
        System.out.println("All employees work for: " + Employee.getCompanyName());
        System.out.println("Changing company name...");
        Employee.setCompanyName("NextGen Technologies");
        System.out.println("Now all employees work for: " + Employee.getCompanyName());
        emp1.displayInfo();  // Will show new company name
        
        // Create more employees
        System.out.println("\n9. Creating more employees:");
        Employee emp4 = new Employee("David Brown", 70000);
        Employee emp5 = new Employee("Eva Martinez", 90000);
        System.out.println("✓ Created 2 more employees");
        
        // Final company info
        System.out.println("\n10. Final company information:");
        Employee.displayCompanyInfo();
        
        System.out.println("\n" + "=".repeat(50));
        System.out.println("TESTING COMPLETE!");
        System.out.println("=".repeat(50));
        
        // IMPORTANT CONCEPTS DEMONSTRATED:
        System.out.println("\n📚 Key Concepts Demonstrated:");
        System.out.println("✓ Static variables (companyName, employeeCount, nextId)");
        System.out.println("✓ Instance variables (name, id, salary)");
        System.out.println("✓ Static methods (setCompanyName, getEmployeeCount, displayCompanyInfo)");
        System.out.println("✓ Instance methods (displayInfo, giveRaise)");
        System.out.println("✓ Automatic ID generation using static variable");
        System.out.println("✓ Shared data across all instances");
        System.out.println("✓ Calling static methods without objects");
    }
}

/*
 * LEARNING NOTES:
 * ===============
 * 
 * STATIC vs INSTANCE:
 * -------------------
 * 
 * STATIC (Class-level):
 * - Belongs to the class itself
 * - Shared by all instances
 * - Can be accessed without creating an object
 * - Memory allocated once when class is loaded
 * - Example: Employee.getEmployeeCount()
 * 
 * INSTANCE (Object-level):
 * - Belongs to each individual object
 * - Each object has its own copy
 * - Requires an object to access
 * - Memory allocated when object is created
 * - Example: emp1.displayInfo()
 * 
 * WHEN TO USE STATIC:
 * -------------------
 * 1. Constants (final static variables)
 *    Example: public static final double PI = 3.14159;
 * 
 * 2. Counters/Shared data
 *    Example: employeeCount, nextId
 * 
 * 3. Utility methods (don't need object state)
 *    Example: Math.sqrt(), Integer.parseInt()
 * 
 * 4. Factory methods
 *    Example: Creating objects in a controlled way
 * 
 * 5. Singleton pattern
 *    Example: Ensuring only one instance exists
 * 
 * WHEN TO USE INSTANCE:
 * ---------------------
 * 1. Object-specific data
 *    Example: employee name, salary
 * 
 * 2. Methods that work with object state
 *    Example: giveRaise() modifies specific employee's salary
 * 
 * 3. Different values for different objects
 *    Example: Each employee has different name/salary
 * 
 * IMPORTANT RULES:
 * ----------------
 * 1. Static methods CANNOT access instance variables directly
 *    - They don't know which object's variables to use
 *    - Must create an object first or receive it as parameter
 * 
 * 2. Static methods CANNOT use 'this' keyword
 *    - 'this' refers to current object
 *    - Static methods don't belong to any object
 * 
 * 3. Instance methods CAN access both static and instance members
 *    - They belong to an object, which belongs to the class
 * 
 * 4. Static variables are initialized before instance variables
 *    - When class is loaded into memory
 * 
 * COMMON MISTAKES:
 * ================
 * 1. Trying to access instance variables from static methods
 *    ✗ public static void test() { System.out.println(name); }
 * 
 * 2. Making everything static when it shouldn't be
 *    - Overusing static defeats the purpose of OOP
 * 
 * 3. Not understanding that static variables are shared
 *    - Changing a static variable affects ALL instances
 * 
 * 4. Forgetting that static methods can be called without objects
 *    - Employee.setCompanyName("X"); // No object needed!
 * 
 * BEST PRACTICES:
 * ===============
 * 1. Use static for truly shared data
 * 2. Use static final for constants
 * 3. Keep most methods instance-based for proper OOP
 * 4. Use static methods for utilities that don't need object state
 * 5. Document when methods are static vs instance
 * 6. Be careful with static mutable state (thread safety)
 */
